lib/pulselistener.js

```diff
@@ -22,56 +24,85 @@ var buildPulseConnectionString = function(options) {
 };
 
+
 /**
  * Create PulseConnection from `options` on the form:
  * {
- *   namespace:         // Namespace to prefix queues/exchanges (optional)
- *                      // defaults to `username` if given otherwise ""
- *   username:          // Username to connect with (and namespace if not given)
- *   password:          // Password to connect with
- *   hostname:          // Hostname to connect to using username/password
- *                      // defaults to pulse.mozilla.org
- *   connectionString:  // connectionString cannot be used with username,
- *                      // password and/or hostname.
+ *   namespace:           // Namespace to prefix queues/exchanges (optional)
+ *                        // defaults to `username` if given otherwise ""
+ *   username:            // Username to connect with (namespace if not given)
+ *   password:            // Password to connect with
+ *   hostname:            // Hostname to connect to using username/password
+ *                        // defaults to pulse.mozilla.org
+ *   protocol:            // 'amqp' or 'amqps', defaults to 'amqps'
+ *   port:                // Port to connect to, defaults to 5671 if 'amqps',
+ *                        // otherwise it defaults to 5672
+ *   connectionString:    // connectionString cannot be used with username,
+ *                        // password and/or hostname.
+ *   reconnect:           // Whether or not to reconnect on errors
+ *   delayFactor:         // Multiplier for: delay = 2 ^ retry * delayFactor
+ *   randomizationFactor: // rf in: delay = delay * random([1 - rf; 1 + rf])
+ *   maxDelay:            // Maximum retry delay
+ *   retryResetTime:      // Time to wait before resetting retry count
+ *   retries:             // Maximum number of retries
  * }
+ *
+ * Note, when starting to reconnect the `reconnect` event is emitted and users
+ * of the connection managed by this object should stop using their connection
+ * and channel objects, and instead call `connect()` and reconfigure themselves
+ * or they should emit and error at their on level if they can't restart.
  */
 var PulseConnection = function(options) {
   assert(typeof(options) === 'object', "options is required");
+  if (options.connectionString) {
+    assert(!options.hostname, "Both `hostname` and `connectionString` given");
+    assert(!options.protocol, "Both `protocol` and `connectionString` given");
+    assert(!options.port,     "Both `port` and `connectionString` given");
+    assert(!options.username, "Both `username` and `connectionString` given");
+    assert(!options.password, "Both `password` and `connectionString` given");
+  }
+
   options = _.defaults({}, options, {
-    namespace:          options.username || ''
+    namespace:            options.username || '',
+    hostname:             'pulse.mozilla.org',
+    protocol:             options.port === 5672 ? 'amqp' : 'amqps',
+    port:                 options.protocol === 'amqp' ? 5672 : 5671,
```

Comment:
`_.defaults({}, options, {port: xzy});` only overwrites `options.port` if no value is given... So non-std port should still work.
 This just tries to guess an appropriate default port.